core/src/network/peer.rs

```diff
@@ -0,0 +1,523 @@
+    > {
+        match self {
+            Peer::Connected(peer) => Some(peer),
+            _ => None,
+        }
+    }
+
+    /// If a connection is pending, returns the `DialingPeer`.
+    pub fn into_dialing(self) -> Option<
+        DialingPeer<'a, TTrans, TInEvent, TOutEvent, THandler, TConnInfo, TPeerId>
+    > {
+        match self {
+            Peer::Dialing(peer) => Some(peer),
+            _ => None,
+        }
+    }
+
+    /// If we are not connected, returns the `DisconnectedPeer`.
+    pub fn into_disconnected(self) -> Option<
+        DisconnectedPeer<'a, TTrans, TInEvent, TOutEvent, THandler, TConnInfo, TPeerId>
+    > {
+        match self {
+            Peer::Disconnected(peer) => Some(peer),
+            _ => None,
+        }
+    }
+}
+
+/// The representation of a peer in a [`Network`] to whom at least
+/// one established connection exists.
+pub struct ConnectedPeer<'a, TTrans, TInEvent, TOutEvent, THandler, TConnInfo, TPeerId>
+where
+    TTrans: Transport,
+    THandler: IntoConnectionHandler<TConnInfo>,
+{
+    network: &'a mut Network<TTrans, TInEvent, TOutEvent, THandler, TConnInfo, TPeerId>,
+    peer_id: TPeerId,
+}
+
+impl<'a, TTrans, TInEvent, TOutEvent, THandler, TConnInfo, TPeerId>
+    ConnectedPeer<'a, TTrans, TInEvent, TOutEvent, THandler, TConnInfo, TPeerId>
+where
+    TTrans: Transport,
+    THandler: IntoConnectionHandler<TConnInfo>,
+    TConnInfo: ConnectionInfo<PeerId = TPeerId>,
+    TPeerId: Eq + Hash + Clone,
+{
+    /// Attempts to establish a new connection to this peer using the given addresses.
+    ///
+    /// Existing established connections are not affected.
+    pub fn connect<I, TMuxer>(self, address: Multiaddr, remaining: I, handler: THandler)
+        -> Result<DialingPeer<'a, TTrans, TInEvent, TOutEvent, THandler, TConnInfo, TPeerId>,
+                  ConnectionLimit>
+    where
+        I: IntoIterator<Item = Multiaddr>,
+        THandler: Send + 'static,
+        THandler::Handler: Send,
+        <THandler::Handler as ConnectionHandler>::Error: error::Error + Send,
+        <THandler::Handler as ConnectionHandler>::OutboundOpenInfo: Send,
+        THandler::Handler: ConnectionHandler<Substream = Substream<TMuxer>, InEvent = TInEvent, OutEvent = TOutEvent> + Send,
+        TTrans: Transport<Output = (TConnInfo, TMuxer)> + Clone,
+        TTrans::Error: Send + 'static,
+        TTrans::Dial: Send + 'static,
+        TMuxer: StreamMuxer + Send + Sync + 'static,
+        TMuxer::OutboundSubstream: Send,
+        TMuxer::Substream: Send,
+        TConnInfo: fmt::Debug + Send + 'static,
+        TPeerId: Eq + Hash + Clone + Send + 'static,
+        TInEvent: Send + 'static,
+        TOutEvent: Send + 'static,
+
+    {
+        if self.network.dialing.contains_key(&self.peer_id) {
+            let mut peer = DialingPeer {
+                network: self.network,
+                peer_id: self.peer_id
+            };
+            peer.connection().add_address(address);
```

Comment:
This behaviour seems a bit weird to me. We add the new addresses to a queue for dialing?
Does that mean that if I call `connect` three times I will still have only one new connection?

Reply:
If we keep the constraint of a single dialing (i.e. pending outgoing) connection per peer, which I have done so far as mentioned in the PR description, this seemed to me to be the natural thing to do, though it should be properly documented, of course. I.e. calling `connect()` on an already `ConnectedPeer` expresses the intention to establish another (new) connection and if such an attempt is already in progress, that same attempt is "picked up", just adding the given addresses as potential further candidates. In other words, if you really want multiple new connections you have to wait for the previous dialing connection to finish.

What alternatives do you see? Returning `Either<PendingPeer, ConnectionLimit>` in the "error" case? Renaming the function? Or changing the semantics in another way? Or would you even prefer if we remove the constraint of a single dialing attempt per peer?

Reply:
I think that in the future we should remove the single dialing attempt per peer (cc https://github.com/libp2p/rust-libp2p/issues/925).
I don't think it should be part of this PR, because I don't think it's an easy task. For instance, we don't want to open 20 TCP connections whenever we connect to a node with 20 addresses.

My complaint here is indeed more about:

- The behaviour of this function being a bit strange.
- The function name not really reflecting its behaviour.

However I don't really have a suggestion, and I'm fine with adding a comment in #925 about this function needing some tweaking.

Reply:
> I think that in the future we should remove the single dialing attempt per peer (cc #925).
I don't think it should be part of this PR, because I don't think it's an easy task. For instance, we don't want to open 20 TCP connections whenever we connect to a node with 20 addresses.

I'm not sure it would be that difficult actually and though I initially thought keeping this constraint would be a good thing, I tend to agree now that it may actually not be desirable in the long run. My naive approach would be that every call to `connect()` would simply result in a new dialing connection for the given addresses. If you have a connection ID of a dialing connection you can still add addresses while dialing is ongoing. I may give it a try and either get back here with a diff or abort if it gets too cumbersome and rather adjust the function name / documentation.

Reply:
[The latest commit](https://github.com/libp2p/rust-libp2p/pull/1440/commits/87452c82d7fff046d8127a0bac47fbd1bcf5422c) shows my current proposal. It essentially establishes more symmetry in the `Peer` API due to:

  1. A `Peer` can `connect()` in any representation, whether disconnected, connected or dialing. Thus there is now `Peer::connect()`.
  2. A `DialingPeer` now has methods `connections()`, `connection(id)` and `num_connections()` analogously to `ConnectedPeer`, which operate on `DialingConnection`s instead of `EstablishedConnection`s.

Reply:
I'm open to extracting that work as a separate, follow-up PR for easier review and sicussion, as the changeset got a bit larger than I anticipated. Let me know.

Reply:
> I'm open to extracting that work as a separate, follow-up PR for easier review and discussion, as the changeset got a bit larger than I anticipated. Let me know.

I'm ok with that.

Reply:
Your original critique w.r.t. this function is now intended to be addressed in this PR by [this commit](https://github.com/libp2p/rust-libp2p/pull/1440/commits/35405880ea281f49cac2fedcf9e444b901a26e5c). Let me know if that is acceptable. I will follow-up with another PR w.r.t. removing the constraint of a single ongoing dialing attempt per peer.